cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -91,6 +95,377 @@ struct CountImpl : public ScalarAggregator {
+
+          int64_t new_size = new_num_groups * sizeof(CType);
+          RETURN_NOT_OK(checked_cast<ResizableBuffer*>(vals)->Resize(new_size));
+
+          auto raw_vals = reinterpret_cast<CType*>(vals->mutable_data());
+          for (int64_t i = old_num_groups; i != new_num_groups; ++i) {
+            raw_vals[i] = pair.second;
+          }
+          return Status::OK();
+        };
+      }
+
+      return Status::OK();
+    }
+
+    Status Visit(const BooleanType& type) {
+      return Status::NotImplemented("Grouped MinMax data of type ", type);
+    }
+
+    Status Visit(const HalfFloatType& type) {
+      return Status::NotImplemented("Grouped MinMax data of type ", type);
+    }
+
+    Status Visit(const DataType& type) {
+      return Status::NotImplemented("Grouped MinMax data of type ", type);
+    }
+
+    ConsumeImpl consume_impl;
+    ResizeImpl resize_min_impl, resize_max_impl;
+  };
+
+  static std::unique_ptr<GroupedMinMaxImpl> Make(
+      KernelContext* ctx, const std::shared_ptr<DataType>& input_type,
+      const FunctionOptions* options) {
+    auto out = ::arrow::internal::make_unique<GroupedMinMaxImpl>();
+    out->options_ = *checked_cast<const MinMaxOptions*>(options);
+    out->type_ = input_type;
+
+    out->buffers_.resize(4);
+    for (auto& buf : out->buffers_) {
+      ctx->SetStatus(ctx->Allocate(0).Value(&buf));
+      if (ctx->HasError()) return nullptr;
+    }
+
+    GetImpl get_impl;
+    ctx->SetStatus(VisitTypeInline(*input_type, &get_impl));
+
+    out->consume_impl_ = std::move(get_impl.consume_impl);
+    out->resize_min_impl_ = std::move(get_impl.resize_min_impl);
+    out->resize_max_impl_ = std::move(get_impl.resize_max_impl);
+    return out;
+  }
+
+  void Resize(KernelContext* ctx, int64_t new_num_groups) override {
+    auto old_num_groups = num_groups_;
+    num_groups_ = new_num_groups;
+
+    KERNEL_RETURN_IF_ERROR(ctx, resize_min_impl_(buffers_[0].get(), new_num_groups));
+    KERNEL_RETURN_IF_ERROR(ctx, resize_max_impl_(buffers_[1].get(), new_num_groups));
+
+    for (auto buffer : {buffers_[2].get(), buffers_[3].get()}) {
+      KERNEL_RETURN_IF_ERROR(ctx, checked_cast<ResizableBuffer*>(buffer)->Resize(
+                                      BitUtil::BytesForBits(new_num_groups)));
+      BitUtil::SetBitsTo(buffer->mutable_data(), old_num_groups, new_num_groups, false);
+    }
+  }
+
+  void Consume(KernelContext* ctx, const Datum& aggregand,
+               const uint32_t* group_ids) override {
+    MaybeResize(ctx, aggregand.length(), group_ids);
+    if (ctx->HasError()) return;
+    consume_impl_(aggregand.array(), group_ids, &buffers_);
+  }
+
+  void Finalize(KernelContext* ctx, Datum* out) override {
+    // aggregation for group is valid if there was at least one value in that group
+    std::shared_ptr<Buffer> null_bitmap = std::move(buffers_[3]);
+
+    if (options_.null_handling == MinMaxOptions::EMIT_NULL) {
+      // ... and there were no nulls in that group
```

Comment:
At the time of writing this I wasn't immediately confident we could use a buffer as both an input and output to UnalignedBitmapOp. After looking through BitmapWordWriter I think this would be safe

Reply:
... I guess it's moot since both these buffer will be aligned.